### DISCUSSION TITLE
Feature Request: named window in Python relational API

### DISCUSSION THREAD

**Discussion by MarcoGorelli** [duckdb/duckdb#14782] [Ideas]
I'm trying to use the Python Relational API, and generally really like it

I have a request for named windows. For example, in SQL I can write:
```python
duckdb.sql("""
SELECT *,
       CASE
           WHEN COUNT(price) OVER w >= 2
           THEN MEAN(price) OVER w
           ELSE NULL
       END AS price_rolling
FROM 'assets.csv'
WHERE symbol in ('ABBV', 'XOM')
WINDOW w AS (
    PARTITION BY symbol
    ORDER BY date
    ROWS BETWEEN 1 PRECEDING AND CURRENT ROW
)
ORDER BY symbol, date
""")
```
and it works wonderfully

With the Python API, I need to write:
```python
window = """
    partition by symbol 
    order by date 
    rows between 1 preceding and current row
"""

(
    duckdb.read_csv("assets.csv")
    .mean(
        "price",
        projected_columns="*",
        window_spec=f"over ({window}) as rolling_price",
    )
    .count(
        "price",
        projected_columns="*",
        window_spec=f"over ({window}) as count_price",
    )
    .select(
      """
      * exclude(rolling_price, count_price),
      case
          when count_price >= 2 then rolling_price
          else null
      end as rolling_price
      """
    )
    .order("symbol, date")
)
```

which is a lot more verbose and - in my opinion - clunky

Is there a better way to write it? If not, might named windows be considered as a feature request, in order to enable less clunky syntax on the Python API side?

**Comment by Tishj** [comment]
I think that ties in nicely to this plan <https://github.com/duckdb/duckdb/discussions/14725#discussioncomment-11200348>
And highlights an option I wasn't aware of, being able to define window specs and reference them by name.
The `window_spec` in the proposed prototype for WindowExpression should probably also be able to accept a `str` to reference an existing window

I think we'll add a `windows` keyword argument to `select`/`project` to be able to define the windows to reference

**Reply by MarcoGorelli** [reply]
amazing, thanks!
